File: Leetcode/0264 Ugly Number II.java
import java.util.*;

class Solution {
    public int nthUglyNumber(int n) {
        int i = 0, uglyNumber = 0;
        var uglyNumbers = new PriorityQueue<>(Comparator.<Integer>naturalOrder().reversed());
        uglyNumbers.add(1);
        while (i < n) {
            uglyNumber = uglyNumbers.poll();
            ++i;
            for (var factor : List.of(2, 3, 5)) {
                int next = factor * uglyNumber;
                if (!uglyNumbers.contains(next)) {
                    uglyNumbers.add(next);
                }
            }
        }
        return uglyNumber;
    }
}

File: Leetcode/0264 Ugly Number II.py
from sortedcontainers import SortedSet

class Solution:
    def nthUglyNumber(self, n):
        ugly_numbers = SortedSet([1])
        i = 0
        while ugly_numbers and i < n:
            ugly_number = ugly_numbers.pop(0)
            i += 1
            for factor in (2, 3, 5):
                ugly_numbers.add(ugly_number * factor)

        return ugly_number